tools/ci_build/op_registration_validator.py

```diff
@@ -0,0 +1,99 @@
+                  # and became an official op.
+                  'kOnnxDomain:MeanVarianceNormalization': 9,
+                  'kOnnxDomain:ThresholdedRelu': 10}
+
+
+class RegistrationValidator(op_registration_utils.RegistrationProcessor):
+    def __init__(self):
+        self.last_op_registrations = {}
+        self.failed = False
+
+    def process_registration(self, lines: typing.List[str], domain: str, operator: str,
+                             start_version: int, end_version: int = None, input_type: str = None):
+        key = domain + ':' + operator
+        prev_start, prev_end = self.last_op_registrations[key] if key in self.last_op_registrations else (None, None)
+
+        if prev_start:
+            # a typed registration where the to/from matches for each entry so nothing to update
+            if prev_start == start_version and prev_end == end_version:
+                return
+
+            # previous registration was unversioned but should have been if we are seeing another registration
+            if not prev_end:
+                log.error("Invalid registration for {}. Registration for opset {} has no end version but was "
+                          "superceeded by version {}."
+                          .format(key, prev_start, start_version))
+                self.failed = True
+                return
+
+            # previous registration end opset is not adjacent to the start of the next registration
+            if prev_end != start_version - 1:
+                log.error("Invalid registration for {}. Registration for opset {} should have end version of {}"
+                          .format(key, prev_start, start_version - 1))
+                self.failed = True
+                return
+
+        self.last_op_registrations[key] = (start_version, end_version)
+
+    def ok(self):
+        return not self.failed
+
+    def validate_last_registrations(self):
+        # make sure we have an unversioned last entry for each operator unless it's deprecated
+        for entry in self.last_op_registrations.items():
+            key, value = entry
+            opset_from, opset_to = value
+
+            deprecated = key in deprecated_ops and opset_to == deprecated_ops[key] - 1
+            if opset_to and not deprecated:
+                log.error('Missing unversioned registration for {}'.format(key))
+                self.failed = True
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(
+        description="Script to validate operator kernel registrations.")
+
+    parser.add_argument(
+        "--ort_root", type=str, help="Path to ONNXRuntime repository root. "
+                                     "Inferred from the location of this script if not provided.")
+
+    args = parser.parse_args()
+
+    ort_root = os.path.abspath(args.ort_root) if args.ort_root else ''
+    include_cuda = True  # validate CPU and CUDA EP registrations
+
```

Comment:
I guess this does not really matter, but it will be nice this can be a param also and default to true

Reply:
Might as well validate all the registrations. Seems an unnecessary complication to have a param to do a subset of files (vs the use case where you're disabling registrations and modifying files where it makes more sense to do that sort of thing).